src/Event/Decorator/BaseDecorator.php

```diff
@@ -0,0 +1,79 @@
+<?php
+/**
+ * CakePHP : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP Project
+ * @since         3.3.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Event\Decorator;
+
+class BaseDecorator
+{
+
+    /**
+     * Callable
+     *
+     * @var callable
+     */
+    protected $_callable;
+
+    /**
+     * Decorator options
+     *
+     * @var array
+     */
+    protected $_options = [];
+
+    /**
+     * Constructor.
+     *
+     * @param callable $callable Callable.
+     * @param array $options Decorator options.
+     */
+    public function __construct(callable $callable, array $options = [])
+    {
+        $this->_callable = $callable;
+        $this->_options = $options;
+    }
+
+    /**
+     * Invoke
+     *
+     * @link http://php.net/manual/en/language.oop5.magic.php#object.invoke
+     * @return mixed
+     */
+    public function __invoke()
+    {
+        return $this->call($this->_callable, func_get_args());
+    }
+
+    /**
+     * Calls a callable with the passed arguments.
+     *
+     * @param callable $callable The callable.
+     * @param array $args Arguments for the callable.
+     * @return mixed
+     */
+    public function call(callable $callable, $args)
```

Comment:
@burzum Seems this method is not used anywhere else. Why it needs to be public if usually the trigger is done via the invoke? It almost seems like an static method that doesn't depend from the rest of the class.

I would suggest you to make this method protected and remove the callable parameter and using the `$this->_callable` directly. On classes that extend this base class it can be called directly without calling the `__invoke`.